src/core/lib/event_engine/windows/grpc_polled_fd_windows.cc

```diff
@@ -0,0 +1,711 @@
+    return;
+  }
+  WSABUF buffer;
+  buffer.buf = reinterpret_cast<char*>(GRPC_SLICE_START_PTR(read_buf_));
+  buffer.len = GRPC_SLICE_LENGTH(read_buf_);
+  memset(winsocket_->read_info()->overlapped(), 0, sizeof(OVERLAPPED));
+  recv_from_source_addr_len_ = sizeof(recv_from_source_addr_);
+  DWORD flags = 0;
+  if (WSARecvFrom(winsocket_->raw_socket(), &buffer, 1, nullptr, &flags,
+                  reinterpret_cast<sockaddr*>(recv_from_source_addr_),
+                  &recv_from_source_addr_len_,
+                  winsocket_->read_info()->overlapped(), nullptr)) {
+    int wsa_last_error = WSAGetLastError();
+    char* msg = gpr_format_message(wsa_last_error);
+    GRPC_ARES_RESOLVER_TRACE_LOG(
+        "fd:|%s| RegisterForOnReadableLocked WSARecvFrom error code:|%d| "
+        "msg:|%s|",
+        GetName(), wsa_last_error, msg);
+    gpr_free(msg);
+    if (wsa_last_error != WSA_IO_PENDING) {
+      ScheduleAndNullReadClosure(GRPC_WSA_ERROR(wsa_last_error, "WSARecvFrom"));
+      return;
+    }
+  }
+  outer_read_closure_ = std::make_unique<AnyInvocableClosure>(
+      [self = shared_from_this()]() mutable {
+        self->OnIocpReadable();
+        self.reset();
+      });
+  winsocket_->NotifyOnRead(outer_read_closure_.get());
+}
+
+void GrpcPolledFdFactoryWindows::GrpcPolledFdWindows::
+    RegisterForOnWriteableLocked(
+        absl::AnyInvocable<void(absl::Status)> write_closure) {
+  if (socket_type_ == SOCK_DGRAM) {
+    GRPC_ARES_RESOLVER_TRACE_LOG("fd:|%s| RegisterForOnWriteableLocked called",
+                                 GetName());
+  } else {
+    GPR_ASSERT(socket_type_ == SOCK_STREAM);
+    GRPC_ARES_RESOLVER_TRACE_LOG(
+        "fd:|%s| RegisterForOnWriteableLocked called tcp_write_state_: %d "
+        "connect_done_: %d",
+        GetName(), tcp_write_state_, connect_done_);
+  }
+  GPR_ASSERT(write_closure_ == nullptr);
+  write_closure_ = std::move(write_closure);
+  if (!connect_done_) {
+    GPR_ASSERT(!pending_continue_register_for_on_writeable_locked_);
+    pending_continue_register_for_on_writeable_locked_ = true;
+    // Register an async OnTcpConnect callback here rather than when the
+    // connect was initiated, since we are now guaranteed to hold a ref of the
+    // c-ares wrapper before write_closure_ is called.
+    winsocket_->NotifyOnWrite(&on_tcp_connect_locked_);
+  } else {
+    ContinueRegisterForOnWriteableLocked();
+  }
+}
+
+void GrpcPolledFdFactoryWindows::GrpcPolledFdWindows::
+    ContinueRegisterForOnWriteableLocked() {
+  GRPC_ARES_RESOLVER_TRACE_LOG(
+      "fd:|%s| ContinueRegisterForOnWriteableLocked "
+      "wsa_connect_error_:%d",
+      GetName(), wsa_connect_error_);
+  GPR_ASSERT(connect_done_);
+  if (wsa_connect_error_ != 0) {
+    ScheduleAndNullWriteClosure(GRPC_WSA_ERROR(wsa_connect_error_, "connect"));
+    return;
+  }
+  if (socket_type_ == SOCK_DGRAM) {
+    ScheduleAndNullWriteClosure(absl::OkStatus());
```

Comment:
you could return after this and eliminate the `else` and one level of nesting.

Reply:
Done.